engine/src/main/java/org/entando/entando/aps/system/services/widgettype/cache/WidgetTypeManagerCacheWrapper.java

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright 2018-Present Entando Inc. (http://www.entando.com) All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+package org.entando.entando.aps.system.services.widgettype.cache;
+
+import com.agiletec.aps.system.exception.ApsSystemException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import org.entando.entando.aps.system.services.widgettype.IWidgetTypeDAO;
+import org.entando.entando.aps.system.services.widgettype.WidgetType;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cache.Cache;
+import org.springframework.cache.CacheManager;
+
+/**
+ * @author E.Santoboni
+ */
+public class WidgetTypeManagerCacheWrapper implements IWidgetTypeManagerCacheWrapper {
+
+	private static final Logger _logger = LoggerFactory.getLogger(WidgetTypeManagerCacheWrapper.class);
+
+	private CacheManager _springCacheManager;
+
+	@Override
+	public void loadWidgetTypes(IWidgetTypeDAO widgetTypeDAO) throws ApsSystemException {
+		try {
+			Cache cache = this.getCache();
+			List<String> codes = (List<String>) this.get(cache, WIDGET_TYPE_CODES_CACHE_NAME, List.class);
+			if (null != codes) {
+				for (int i = 0; i < codes.size(); i++) {
+					String code = codes.get(i);
+					cache.evict(WIDGET_TYPE_CACHE_NAME_PREFIX + code);
+				}
+				cache.evict(WIDGET_TYPE_CODES_CACHE_NAME);
+			}
+			Map<String, WidgetType> widgetTypes = widgetTypeDAO.loadWidgetTypes();
+			Iterator<WidgetType> iter1 = widgetTypes.values().iterator();
+			while (iter1.hasNext()) {
+				WidgetType type = iter1.next();
+				String mainTypeCode = type.getParentTypeCode();
+				if (null != mainTypeCode) {
+					type.setParentType(widgetTypes.get(mainTypeCode));
+				}
+			}
+			List<String> widgetCodes = new ArrayList<String>();
+			Iterator<WidgetType> iter2 = widgetTypes.values().iterator();
+			while (iter2.hasNext()) {
+				WidgetType type = iter2.next();
+				cache.put(WIDGET_TYPE_CACHE_NAME_PREFIX + type.getCode(), type);
+				widgetCodes.add(type.getCode());
+			}
+			cache.put(WIDGET_TYPE_CODES_CACHE_NAME, widgetCodes);
+		} catch (Throwable t) {
+			_logger.error("Error loading widgets types", t);
+			throw new ApsSystemException("Error loading widgets types", t);
+		}
+	}
+
+	@Override
+	public WidgetType getWidgetType(String code) {
+		return this.get(WIDGET_TYPE_CACHE_NAME_PREFIX + code, WidgetType.class);

```

Comment:
Is this call read through? If its a cache miss does it check the backing store or does the caller have to handle that?

Reply:
If no cached object is found, the method returns null; no call is made to the underlying layer